osquery/tables/system/windows/drivers.cpp

```diff
@@ -173,79 +144,135 @@ Status getDeviceProperty(const device_infoset_t& infoset,
-        r["driver_key"].insert(0, kDriverKeyPath);
-      }
+    if (r.count("driver_key") > 0 && !r.at("driver_key").empty()) {
+      r["driver_key"].insert(0, kDriverKeyPath);
+    }
+
+    if (r.count("service") > 0 && !r.at("service").empty()) {
+      std::string svcKey = kServiceKeyPath + r["service"];
+      r["service_key"] = svcKey;
+      r["image"] = getDriverImagePath(svcKey);
     }
-    if (r.count("service") > 0) {
-      if (!r.at("service").empty()) {
-        r["service_key"] = kServiceKeyPath + r["service"];
-        r["image"] = getDriverImagePath(r["service_key"]);
-      }
+    apiDevices[devId] = r;
+  }
+
+  WmiRequest wmiSignedDriverReq("select * from Win32_PnPSignedDriver");
+  auto& wmiResults = wmiSignedDriverReq.results();
+
+  if (wmiResults.empty()) {
+    LOG(WARNING) << "Failed to query device drivers via WMI";
+    return {};
+  }
+
+  /*
+   * We balance getting information from WMI and Win32 APIs as not
+   * all data is available in only one place. Unfortunately this means
+   * two runs through the devices list, but this takes less time
+   * than the Win32 API method
+   */
+  for (const auto& row : wmiResults) {
+    Row r;
+    std::string devid;
+    row.GetString("DeviceID", devid);
+    r["device_id"] = devid;
+    row.GetString("DeviceName", r["device_name"]);
+    row.GetString("Description", r["description"]);
+    row.GetString("DeviceClass", r["class"]);
+    row.GetString("DriverVersion", r["version"]);
+    row.GetString("Manufacturer", r["manufacturer"]);
+    row.GetString("DriverProviderName", r["provider"]);
+
+    bool isSigned;
+    row.GetBool("IsSigned", isSigned);
+    r["signed"] = isSigned ? INTEGER(1) : INTEGER(0);
+
+    std::string infName;
+    row.GetString("InfName", infName);
+    std::vector<char> inf(MAX_PATH, 0x0);
+    unsigned long infLen = 0;
+    auto sdiRet =
+        SetupGetInfDriverStoreLocation(infName.c_str(),
+                                       nullptr,
+                                       nullptr,
+                                       inf.data(),
+                                       static_cast<unsigned long>(inf.size()),
+                                       &infLen);
+    if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+      inf.resize(infLen);
+      sdiRet =
+          SetupGetInfDriverStoreLocation(infName.c_str(),
+                                         nullptr,
+                                         nullptr,
+                                         inf.data(),
+                                         static_cast<unsigned long>(inf.size()),
+                                         &infLen);
+    }
+    if (sdiRet != TRUE) {
+      VLOG(1) << "Failed to derive full driver INF path with "
+              << GetLastError();
+      r["inf"] = infName;
+    } else {
+      r["inf"] = inf.data();
+    }
+
+    // Add the remaining columns from the APIs
+    if (apiDevices.find(devid) != apiDevices.end()) {
```

Comment:
Please use the returned iterator here to avoid all the searches on the map with `[]` accesses.

Reply:
This might be a bit tricky. The reason for doing it this way, is that we don't have guarantees on the relationship between the WMI and API derived sets. The WMI set by far has more information and more drivers listed, thus that's what we primarily build this table out of, however it was worth ensuring the device existed in the sets listed by Win32 APIs before attempting to get the information for that device id. There might be a better way to construct this using `std::set`s though and possibly get a boost on lookup speed?